sklearn/neural_network/multilayer_perceptron.py

```diff
@@ -0,0 +1,1233 @@
+
+                    # update weights
+                    grads = coef_grads + intercept_grads
+                    self._optimizer.update_params(grads)
+
+                self.n_iter_ += 1
+
+                self.t_ += n_samples
+                self.loss_curve_.append(self.loss_)
+                if self.verbose:
+                    print("Iteration %d, loss = %.8f" % (self.n_iter_,
+                                                         self.loss_))
+
+                # adjusting learning rates
+                if self.learning_rate == 'invscaling':
+                    self._optimizer.learning_rate = (
+                        self.learning_rate_init / (self.t_ + 1)**self.power_t)
+                # validation set evaluation
+                if early_stopping:
+                    # compute validation score, use that for stopping
+                    self.validation_scores_.append(self.score(X_val, y_val))
+
+                    if self.verbose:
+                        print("Validation score: %f" %
+                              (self.validation_scores_[-1]))
+                    # update best parameters
+                    # use validation_scores_, not loss_curve_
+                    # let's hope no-one overloads .score with mse
+                    last_valid_score = self.validation_scores_[-1]
+
+                    if last_valid_score < (self.best_validation_score_ +
+                                           self.tol):
+                        self._no_improvement_count += 1
+                    else:
+                        self._no_improvement_count = 0
+
+                    if last_valid_score > self.best_validation_score_:
+                        self.best_validation_score_ = last_valid_score
+                        self._best_coefs = [c.copy() for c in self.coefs_]
+                        self._best_intercepts = [i.copy()
+                                                 for i in self.intercepts_]
+                else:
+                    if self.loss_curve_[-1] > self.best_loss_ - self.tol:
+                        self._no_improvement_count += 1
+                    else:
+                        self._no_improvement_count = 0
+                    if self.loss_curve_[-1] < self.best_loss_:
+                        self.best_loss_ = self.loss_curve_[-1]
+
+                # stopping criteria
+                if self._no_improvement_count > 2:
+                    # not better than last two iterations by tol.
+                    # stop or decreate learning rate
+                    if early_stopping:
+                        msg = ("Validation score did not improve more than "
+                               "tol=%f for two consecutive epochs." % self.tol)
+                    else:
+                        msg = ("Training loss did not improve more than tol=%f"
+                               " for two consecutive epochs." % self.tol)
+
+                    if self.learning_rate == 'adaptive':
```

Comment:
Maybe. Does it make sense to do this with adam? Adam uses already decreasing step-sizes, right?